File: quiver2.py
import numpy as np

# FIXME: insertions/delections can cause wrong answers in
# `updated_col()` and functions that rely on it, because they change
# the shape of the matrix.

# TODO: cache BandedMatrix calls, or precompute ranges, if they take too much time
# TODO: bounds checks in BandedMatrix
# TODO: test BandedMatrix
# TODO: improve slicing and arithmetic operations of BandedMatrix
# TODO: port to Julia and compare speed
# TODO: stochastic gradient descent. weight choices by sequence quality.

# TODO: use faster version of partial order aligner
# TODO: benchmark against actual quiver implementation


class BandedMatrix:
    """A sparse banded matrix.

    Band is always wide enough to include [0, 0] and [-1, -1].

    Currently only supports limited slicing and no arithmetic
    operations.

    Warning: does not do bounds checks.

    """

    def __init__(self, shape, bandwidth=0):
        nrows, ncols = shape
        if not nrows > 0 and ncols > 0:
            raise Exception('shape is not positive: {}'.format(shape))
        self.shape = shape
        self.bandwidth = bandwidth
        self.diff = np.abs(nrows - ncols)
        data_nrows = 2 * bandwidth + self.diff + 1
        self.data = np.zeros((data_nrows, ncols))
        self.h_offset = max(ncols - nrows, 0)
        self.v_offset = max(nrows - ncols, 0)

    @property
    def nrows(self):
        return self.shape[0]

    @property
    def ncols(self):
        return self.shape[1]

    def data_row(self, i, j):
        """Returns k where self.data[k, j] stores M[i, j]"""
        return (i - j) + self.h_offset + self.bandwidth

    def __getitem__(self, key):
        i, j = key
        while i < 0:
            i = self.nrows + i
        while j < 0:
            j = self.ncols + j
        row = self.data_row(i, j)
        return self.data[row, j]

    def row_range(self, j):
        """[start, stop) of band in column j"""
        start = max(0, j - self.h_offset - self.bandwidth)
        stop = min(j + self.v_offset + self.bandwidth + 1, self.nrows)
        return start, stop

    def data_row_range(self, j):
        """[start, stop) of band in self.data column j"""
        start = max(0, self.h_offset + self.bandwidth - j)
        a, b = self.row_range(j)
        stop = start + (b - a)
        return start, stop

    def get_col(self, j):
        """banded part only of column j"""
        dstart, dstop = self.data_row_range(j)
        return self.data[dstart:dstop, j]

    def __setitem__(self, key, val):
        i, j = key
        row = self.data_row(i, j)
        self.data[row, j] = val

    def inband(self, i, j):
        """whether position [i, j] is in the band"""
        row = self.data_row(i, j)
        return 0 <= row < self.data.shape[0]

    def full(self):
        """Construct a dense np.array of this matrix."""
        result = np.zeros((self.nrows, self.ncols))
        for j in range(self.ncols):
            start, stop = self.row_range(j)
            dstart, dstop = self.data_row_range(j)
            result[start:stop, j] = self.data[dstart:dstop, j]
        return result

    def flip(self):
        self.data = np.flipud(np.fliplr(self.data))


def update(arr, i, j, s_base, t_base, log_p, log_ins, log_del):
    """compute alignment score of [i, j]"""
    sub = (0 if s_base == t_base else log_p)
    scores = []
    if arr.inband(i - 1, j):
        # insertion
        scores.append(arr[i - 1, j] + log_ins)
    if arr.inband(i, j - 1):
        # deletion
        scores.append(arr[i, j - 1] + log_del)
    if arr.inband(i - 1, j - 1):
        scores.append(arr[i - 1, j - 1] + sub)
    return max(scores)


def forward(s, log_p, t, log_ins, log_del, bandwidth):
    """Forward matrix for aligning `s` to `t`."""
    result = BandedMatrix((len(s) + 1, len(t) + 1), bandwidth)
    for i in range(min(result.shape[0], result.v_offset + bandwidth + 1)):
        result[i, 0] = log_ins * i
    for j in range(min(result.shape[1], result.h_offset + bandwidth + 1)):
        result[0, j] = log_del * j
    # TODO: reverse order of iteration?
    for j in range(1, result.ncols):
        start, stop = result.row_range(j)
        start = max(start, 1)
        for i in range(start, stop):
            result[i, j] = update(result, i, j, s[i-1], t[j-1], log_p[i-1], log_ins, log_del)
    return result


def backward(s, log_p, t, log_ins, log_del, bandwidth):
    """Backward matrix for aligning `s` to `t`."""
    s = ''.join(reversed(s))
    log_p = np.array(list(reversed(log_p)))
    t = ''.join(reversed(t))
    result = forward(s, log_p, t, log_ins, log_del, bandwidth)
    result.flip()
    return result


def updated_col(prev_col, new_col, base, template, seq, log_p, A, log_ins, log_del):
    """Compute a new column for `base` in a particular position.

    prev_col: column from which to calculate new result

    new_col: used only for shape. If this is a substitution, it is
             prev_col + 1; if it is a deletion it is the same as
             prev_col.

    """
    prev = A.get_col(prev_col)
    prev_start, prev_stop = A.row_range(prev_col)
    row_start, row_stop = A.row_range(new_col)
    offset = 1 - (row_start - prev_start)
    result = np.empty((row_stop - row_start))
    for i, real_i in enumerate(range(row_start, row_stop)):
        scores = []
        if i > 0:
            # insertion
            scores.append(result[i - 1] + log_ins)
        if real_i > prev_start and real_i <= prev_stop:
            # match / substitution
            scores.append(prev[i - offset] + (0 if base == seq[real_i-1] else log_p[real_i-1]))
        if real_i >= prev_start and real_i < prev_stop:
            # deletion
            scores.append(prev[i - offset + 1] + log_del)
        result[i] = max(scores)
    return result


def score_mutation(mutation, template, seq, log_p, A, B, log_ins, log_del, bandwidth):
    """Score a mutation using the forward-backward trick."""
    mtype, pos, base = mutation
    bj = pos + (0 if mtype == 'insertion' else 1)
    Bcol = B.get_col(bj)
    if mtype == 'deletion':
        aj = pos
        Acol = A.get_col(aj)
        # need to chop off beginning of Acol and end of Bcol and align
        a_start, a_stop = A.row_range(aj)
        b_start, b_stop = B.row_range(bj)
        amin = max(b_start - a_start, 0)
        amax = len(Acol) - max(a_stop - b_stop, 0)
        bmin = max(a_start - b_start, 0)
        bmax = len(Bcol) - max(b_stop - a_stop, 0)
        return (Acol[amin:amax] + Bcol[bmin:bmax]).max()
    aj = pos + (1 if mtype == 'substitution' else 0)
    Acol = updated_col(pos, aj, base, template, seq, log_p, A, log_ins, log_del)
    return (Acol + Bcol).max()


def mutations(template):
    """Yields all mutations as tuple (name, position, base)"""
    for j in range(len(template)):
        for base in 'ACGT':
            if template[j] == base:
                continue
            yield ['substitution', j, base]
        yield ['deletion', j, None]
        for base in 'ACGT':
            yield ['insertion', j, base]
    # insertion after last
    for base in 'ACGT':
        yield ['insertion', len(template), base]


def update_template(template, mutation):
    """Apply a mutation to a template"""
    mtype, pos, base = mutation
    if mtype == 'substitution':
        return ''.join([template[:pos], base, template[pos + 1:]])
    elif mtype == 'insertion':
        return ''.join([template[:pos], base, template[pos:]])
    elif mtype == 'deletion':
        return ''.join([template[:pos], template[pos + 1:]])
    else:
        raise Exception('unknown mutation: {}'.format(mtype))


def apply_mutations(template, mutations):
    """Apply multiple mutations to a template.

    Keeps track of changes in positions due to insertions and
    deletions.

    """
    for i in range(len(mutations)):
        score, mutation = mutations[i]
        template = update_template(template, mutation)
        mtype, pos, _ = mutation
        if mtype == 'insertion':
            for ii in range(i + 1, len(mutations)):
                if mutations[ii][1][1] > pos:
                    mutations[ii][1][1] += 1
        elif mtype == 'deletion':
            for ii in range(i + 1, len(mutations)):
                if mutations[ii][1][1] > pos:
                    mutations[ii][1][1] -= 1
    return template


def choose_candidates(candidates, min_dist):
    """Choose best mutations that are at least `min_dist` bases distant."""
    final_cands = []
    posns = set()
    for c in reversed(sorted(candidates, key=lambda c: c[0])):
        _, (_, posn, _) = c
        if any(abs(posn - p) < min_dist for p in posns):
            continue
        posns.add(posn)
        final_cands.append(c)
    return final_cands


def quiver2(template, sequences, phreds, log_ins, log_del, bandwidth=10,
            min_dist=9, max_iters=100, verbose=False):
    """Generate an alignment-free consensus.

    sequences: list of dna sequences

    phreds: list of arrays of phred scores

    log_ins: penalty for insertion

    log_del: penalty for deletion

    bandwidth: extra bandwidth in banded alignment matrices

    min_dist: minimum distance between multiple mutations accepted in
              a single iteration

    max_iters: maximum iterations before termination

    verbose: print progress to the command line

    """
    log_ps = list((-phred / 10) for phred in phreds)

    if bandwidth < 0:
        raise Exception('bandwidth cannot be negative: {} '.format(bandwidth))

    if verbose:
        print("computing initial alignments")
    As = list(forward(s, p, template, log_ins, log_del, bandwidth) for s, p in zip(sequences, log_ps))
    Bs = list(backward(s, p, template, log_ins, log_del, bandwidth) for s, p in zip(sequences, log_ps))
    best_score = sum(A[-1, -1] for A in As)
    best_template = template
    current_score = best_score
    current_template = best_template
    if verbose:
        print('initial score: {:.4f}'.format(best_score))
    converged = False
    total_mutations = 0
    for i in range(max_iters):
        old_template = current_template
        old_score = current_score
        if verbose:
            print("iteration {}".format(i))
        candidates = []
        for mutation in mutations(current_template):
            score = sum(score_mutation(mutation, current_template, seq, log_p, A, B, log_ins, log_del, bandwidth)
                        for seq, log_p, A, B in zip(sequences, log_ps, As, Bs))
            if score > current_score:
                candidates.append([score, mutation])
        if not candidates:
            converged = True
            break
        if verbose:
            print("  found {} candidate mutations".format(len(candidates)))
        chosen_cands = choose_candidates(candidates, min_dist)
        if verbose:
            print('  filtered to {} candidate mutations'.format(len(chosen_cands)))
        current_template = apply_mutations(current_template, chosen_cands)
        As = list(forward(s, p, current_template, log_ins, log_del, bandwidth) for s, p in zip(sequences, log_ps))
        Bs = list(backward(s, p, current_template, log_ins, log_del, bandwidth) for s, p in zip(sequences, log_ps))
        current_score = sum(A[-1, -1] for A in As)

        # detect if a single mutation is better
        # FIXME: code duplication
        # FIXME: this may not actually work, because score_mutation() is not exact
        if current_score < chosen_cands[0][0]:
            if verbose:
                print('  rejecting multiple candidates in favor of best')
            chosen_cands = [chosen_cands[0]]
            current_template = apply_mutations(old_template, chosen_cands)
            As = list(forward(s, p, current_template, log_ins, log_del, bandwidth) for s, p in zip(sequences, log_ps))
            Bs = list(backward(s, p, current_template, log_ins, log_del, bandwidth) for s, p in zip(sequences, log_ps))
            current_score = sum(A[-1, -1] for A in As)
        total_mutations += len(chosen_cands)
        if verbose:
            print('  score: {:.4f}'.format(current_score))
        if current_score > best_score:
            best_template = current_template
            best_score = current_score
        if old_score > current_score:
            if verbose:
                # this can happen because score_mutation() is not exact for insertions and deletions
                # FIXME: detect if this keeps happening and return best overall template
                print('Warning: score decreased.')
    info = {
        'converged': converged,
        'iterations': i,
        'mutations': total_mutations,
    }
    return best_template, info

File: quiver2.jl
module Quiver2

include("BandedArray.jl")

using BandedArrayModule

export quiver2

function update(A::BandedArray{Float64}, i::Int, j::Int, s_base::Char, t_base::Char,
                log_p::Float64, log_ins::Float64, log_del::Float64)
    score = typemin(Float64)
    if inband(A, i - 1, j)
        # insertion
        score = max(score, A[i - 1, j] + log_ins)
    end
    if inband(A, i, j - 1)
        # deletion
        score = max(score, A[i, j - 1] + log_del)
    end
    if inband(A, i - 1, j - 1)
        score = max(score, A[i - 1, j - 1] +( s_base == t_base ? 0.0 : log_p))
    end
    return score
end

function forward(s::AbstractString, log_p::Array{Float64, 1}, t::AbstractString,
                 log_ins::Float64, log_del::Float64, bandwidth::Int)
    result = BandedArray(Float64, (length(s) + 1, length(t) + 1), bandwidth)
    for i = 2:min(size(result)[1], result.v_offset + bandwidth + 1)
        result[i, 1] = log_ins * (i - 1)
    end
    for j = 2:min(size(result)[2], result.h_offset + bandwidth + 1)
        result[1, j] = log_del * (j - 1)
    end
    for j = 2:size(result)[2]
        start, stop = row_range(result, j)
        start = max(start, 2)
        for i = start:stop
            result[i, j] = update(result, i, j, s[i-1], t[j-1], log_p[i-1], log_ins, log_del)
        end
    end
    return result
end

function backward(s::AbstractString, log_p::Array{Float64, 1}, t::AbstractString,
                 log_ins::Float64, log_del::Float64, bandwidth::Int)
    s = reverse(s)
    log_p = flipdim(log_p, 1)
    t = reverse(t)
    result = forward(s, log_p, t, log_ins, log_del, bandwidth)
    return flip(result)
end

abstract Mutation

immutable Substitution <: Mutation
    pos::Int
    base::Char
end

immutable Insertion <: Mutation
    pos::Int
    base::Char
end

immutable Deletion <: Mutation
    pos::Int
end

immutable CandMutation
    mutation::Mutation
    score::Float64
end

function updated_col(mutation::Mutation,
                     template::AbstractString, seq::AbstractString,
                     log_p::Array{Float64, 1}, A::BandedArray{Float64},
                     log_ins::Float64, log_del::Float64)
    aj = mutation.pos + (typeof(mutation) == Substitution ? 1 : 0)
    prev::SubArray{Float64,1,Array{Float64,2},Tuple{UnitRange{Int64},Int64},2} = sparsecol(A, mutation.pos)
    prev_start, prev_stop = row_range(A, mutation.pos)
    row_start, row_stop = row_range(A, aj)
    offset = 1 - (row_start - prev_start)
    result = Array{Float64}(row_stop - row_start + 1)
    for real_i in row_start:row_stop
        seq_i = real_i - 1
        i = real_i - row_start + 1
        ii = i - offset + 1
        score = typemin(Float64)
        if i > 1
            score = max(score, result[i - 1] + log_ins)
        end
        if prev_start < real_i <= (prev_stop + 1)
            score = max(score, prev[ii - 1] + (mutation.base == seq[seq_i] ? 0.0 : log_p[seq_i]))
        end
        if prev_start <= real_i <= prev_stop
            score = max(score, prev[ii] + log_del)
        end
        result[i] = score
    end
    return result
end

function equal_ranges(a_range::Tuple{Int64,Int64}, b_range::Tuple{Int64,Int64})
    a_start, a_stop = a_range
    b_start, b_stop = b_range
    alen = a_stop - a_start + 1
    blen = b_stop - b_start + 1
    amin = max(b_start - a_start + 1, 1)
    amax = alen - max(a_stop - b_stop, 0)
    bmin = max(a_start - b_start + 1, 1)
    bmax = blen - max(b_stop - a_stop, 0)
    return (amin, amax), (bmin, bmax)
end

@generated function summax(a, b)
    return quote
        result::Float64 = a[1] + b[1]
        for i = 2:min(length(a), length(b))
            result = max(result, a[i] + b[i])
        end
        return result
    end
end

function score_mutation(mutation::Deletion, template::AbstractString,
                        seq::AbstractString, log_p::Array{Float64, 1},
                        A::BandedArray{Float64}, B::BandedArray{Float64},
                        log_ins::Float64, log_del::Float64, bandwidth::Int)
    aj = mutation.pos
    bj = mutation.pos + 1
    Acol = sparsecol(A, aj)
    Bcol = sparsecol(B, bj)
    (amin, amax), (bmin, bmax) = equal_ranges(row_range(A, aj), row_range(B, bj))

    asub = sub(Acol, amin:amax)
    bsub = sub(Bcol, bmin:bmax)

    return summax(asub, bsub)
end

function score_mutation(mutation::Union{Insertion,Substitution}, template::AbstractString,
                        seq::AbstractString, log_p::Array{Float64, 1},
                        A::BandedArray{Float64}, B::BandedArray{Float64},
                        log_ins::Float64, log_del::Float64, bandwidth::Int)
    Acol::Array{Float64} = updated_col(mutation, template, seq, log_p, A, log_ins, log_del)
    bj = mutation.pos + (typeof(mutation) == Insertion ? 0 : 1)
    Bcol::SubArray{Float64,1,Array{Float64,2},Tuple{UnitRange{Int64},Int64},2} = sparsecol(B, bj)

    return summax(Acol, Bcol)
end

function update_template(template::AbstractString, mutation::Substitution)
    return string(template[1:(mutation.pos - 1)], mutation.base, template[(mutation.pos + 1):end])
end

function update_template(template::AbstractString, mutation::Insertion)
    return string(template[1:(mutation.pos - 1)], mutation.base, template[(mutation.pos):end])
end

function update_template(template::AbstractString, mutation::Deletion)
    return string(template[1:(mutation.pos - 1)], template[(mutation.pos + 1):end])
end

function apply_mutations(template::AbstractString, mutations::Array{Mutation, 1})
    # check that mutations all have different positions.
    # this is a bit too strict, since there are some combinations of
    # mutations affecting the same spot that are unambiguous.
    if length(Set([m.pos for m in mutations])) != length(mutations)
        error("Cannot have multiple mutations affecting the same position")
    end
    remaining = [m for m in mutations]

    while length(remaining) > 0
        m = pop!(remaining)
        template = update_template(template, m)
        o = Dict(Insertion => 1, Deletion => -1, Substitution => 0)[typeof(m)]
        for i in 1:length(remaining)
            m2 = remaining[i]
            if m2.pos >= m.pos
                T = typeof(m2)
                if T == Deletion
                    remaining[i] = T(m2.pos + o)
                else
                    remaining[i] = T(m2.pos + o, m2.base)
                end
            end
        end
    end

    return template
end

function choose_candidates(candidates::Vector{CandMutation}, min_dist::Int)
    final_cands = CandMutation[]
    posns = Set()
    for c in sort(candidates, by=(c) -> c.score, rev=true)
        if any(Bool[(abs(c.mutation.pos - p) < min_dist) for p in posns])
            continue
        end
        push!(posns, c.mutation.pos)
        push!(final_cands, c)
    end
    return final_cands
end

# This macro is used in getcands() for performance reasons, to avoid
# using a variable of type `Mutation`.
#
# TODO: convert this to a generated function.
macro process_mutation(m)
    return quote
        score = 0.0
        for si in 1:length(sequences)
            score += score_mutation($m, template, sequences[si], log_ps[si],
                                    As[si], Bs[si], log_ins, log_del, bandwidth)
        end
        if score > current_score
            push!(candidates, CandMutation($m, score))
        end
    end
end

function getcands(template::AbstractString, current_score::Float64,
                  sequences::Vector{ASCIIString},
                  log_ps::Vector{Vector{Float64}},
                  As::Vector{BandedArray{Float64}}, Bs::Vector{BandedArray{Float64}},
                  log_ins::Float64, log_del::Float64, bandwidth::Int)
    candidates = CandMutation[]
    for j in 1:length(template)
        for base in "ACGT"
            mi = Insertion(j, base)
            @process_mutation mi
            if template[j] == base
                continue
            end
            ms = Substitution(j, base)
            @process_mutation ms
        end
        md = Deletion(j)
        @process_mutation md
    end
    # insertion after last
    for base in "ACGT"
        mi = Insertion(length(template) + 1, base)
        @process_mutation mi
    end
    return candidates
end

function quiver2(template::AbstractString, sequences::Vector{ASCIIString},
                 phreds::Vector{Vector{Float64}},
                 log_ins::Float64, log_del::Float64;
                 bandwidth::Int=10, min_dist::Int=9, batch::Int=10,
                 max_iters::Int=100, verbose::Bool=false)
    log_ps = [(-phred / 10.0) for phred in phreds]

    if bandwidth < 0
        error("bandwidth cannot be negative: $bandwidth")
    end

    if verbose
        print("computing initial alignments\n")
    end

    if batch < 0
        batch = length(sequences)
    end
    batch = min(batch, length(sequences))
    if batch < length(sequences)
        indices = rand(1:length(sequences), batch)
        seqs = sequences[indices]
        lps = log_ps[indices]
    else
        seqs = sequences
        lps = log_ps
    end

    As = [forward(s, p, template, log_ins, log_del, bandwidth) for (s, p) in zip(seqs, lps)]
    Bs = [backward(s, p, template, log_ins, log_del, bandwidth) for (s, p) in zip(seqs, lps)]
    best_score = sum([A[end, end] for A in As])
    best_template = template
    current_score = best_score
    current_template = best_template
    if verbose
        print("initial score: $best_score\n")
    end
    converged = false
    total_mutations = 0
    iterations = 0
    for i in 1:max_iters
        iterations = i
        old_template = current_template
        old_score = current_score
        if verbose
            print("iteration $i\n")
        end

        candidates = getcands(current_template, current_score, seqs, lps, As, Bs, log_ins, log_del, bandwidth)
        if length(candidates) == 0
            if batch < length(sequences)
                if verbose
                    print("no candidates found. switching off batch mode.\n")
                end
                # start full runs
                batch = length(sequences)
                # TODO: this is necessary because unbatched scores will be worse.
                # what's the alternative?
                best_score = typemin(Float64)
                # TODO: instead of turning off batch mode, try increasing batch size
                # TODO: is there some fast way to detect convergence without a full run?
                # TODO: try multiple iterations before changing/disabling batch
            else
                converged = true
                break
            end
        end
        if length(candidates) > 0
            if verbose
                print("  found $(length(candidates)) candidate mutations.\n")
            end
            chosen_cands = choose_candidates(candidates, min_dist)
            if verbose
                print("  filtered to $(length(chosen_cands)) candidate mutations\n")
            end
            current_template = apply_mutations(current_template, Mutation[c.mutation for c in chosen_cands])
            current_score = 0.0
            for i = 1:length(seqs)
                current_score += forward(seqs[i], lps[i], current_template, log_ins, log_del, bandwidth)[end, end]
            end

            # detect if a single mutation is better
            # FIXME: code duplication
            # FIXME: this may not actually work, because score_mutation() is not exact
            if current_score < chosen_cands[1].score
                if verbose
                    print("  rejecting multiple candidates in favor of best\n")
                end
                chosen_cands = CandMutation[chosen_cands[1]]
                current_template = apply_mutations(old_template, Mutation[c.mutation for c in chosen_cands])
                current_score = 0.0
                for i = 1:length(seqs)
                    current_score += forward(seqs[i], lps[i], current_template, log_ins, log_del, bandwidth)[end, end]
                end
            end
            total_mutations += length(chosen_cands)
            if verbose
                print("  score: $current_score\n")
            end
            if current_score > best_score
                best_template = current_template
                best_score = current_score
            end
            if old_score > current_score
                if verbose
                    # this can happen because score_mutation() is not exact for insertions and deletions
                    # FIXME: detect if this keeps happening and return best overall template
                    print("Warning: score decreased.\n")
                end
            end
        end
        if batch < length(sequences)
            indices = rand(1:length(sequences), batch)
            seqs = sequences[indices]
            lps = log_ps[indices]
        else
            seqs = sequences
            lps = log_ps
        end
        As = [forward(s, p, current_template, log_ins, log_del, bandwidth) for (s, p) in zip(seqs, lps)]
        Bs = [backward(s, p, current_template, log_ins, log_del, bandwidth) for (s, p) in zip(seqs, lps)]
        current_score = 0.0
        for i = 1:length(seqs)
            current_score += As[i][end, end]
        end

    end
    info = Dict("converged" => converged,
                "iterations" => iterations,
                "mutations" => total_mutations,
                )
    return best_template, info
end

end
